{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8828f6cf",
   "metadata": {},
   "source": [
    "# Creating transcripts from Xenium Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87aef1c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pixel_value = [(255, 0 ,0), (0, 255, 0)] \n",
    "list_of_genes=['UCHL1', 'FABP7'] \n",
    "genes_to_plot = {}\n",
    "for i in range(len(pixel_value)):\n",
    "    genes_to_plot[list_of_genes[i]]= pixel_value[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6b2c3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run for DRG1\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageDraw\n",
    "import tifffile\n",
    "#put in the input folder for the tissue data and the output path to save the transcript image\n",
    "input_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/Xenium_Data/output-XETG00171__0018220__Region_1__20240207__003742/'\n",
    "output_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_1__20240207__003742.tif'\n",
    "\n",
    "mipurl = input_path + '/morphology_mip.ome.tif'\n",
    "\n",
    "# Assuming `input_path`, `output_path`, `factor`, and `image_size_original` are defined earlier in your script\n",
    "data_file_path = input_path + 'transcripts.csv'  # Ensure `input_path` ends with a slash (/)\n",
    "output_image_path = output_path  # Ensure this variable is correctly defined\n",
    "factor = 0.2125  # Scaling factor to convert pixel into micrometer\n",
    "with tifffile.TiffFile(mipurl) as tif:\n",
    "    image_size_original=tif.series[0].levels[0].shape\n",
    "\n",
    "# Adjust `image_size_original` function or variable appropriately to get original image size\n",
    "image_size = (int(image_size_original[1]*factor), int(image_size_original[0]*factor))\n",
    "\n",
    "\n",
    "# Read the spatial sequencing data\n",
    "data = pd.read_csv(data_file_path)\n",
    "\n",
    "# Create a black image\n",
    "image = Image.new('RGB', image_size, 'black')\n",
    "draw = ImageDraw.Draw(image)\n",
    "\n",
    "# Scale the x_location and y_location by 'factor'\n",
    "data['x_location_scaled'] = data['x_location'] \n",
    "data['y_location_scaled'] = data['y_location'] \n",
    "\n",
    "# Plot the points for the specified genes with qv > 20\n",
    "for index, row in data.iterrows():\n",
    "    if row['qv'] > 20 and row['feature_name'] in genes_to_plot:\n",
    "        color = genes_to_plot[row['feature_name']]\n",
    "        # Plot the gene\n",
    "        draw.point((row['x_location_scaled'], row['y_location_scaled']), fill=color)\n",
    "\n",
    "# Save the image\n",
    "image.save(output_image_path, format='TIFF')\n",
    "\n",
    "print(f\"Image saved to {output_image_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ae136e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run for DRG2\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageDraw\n",
    "import tifffile\n",
    "#put in the input folder for the tissue data and the output path to save the transcript image\n",
    "input_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/Xenium_Data/output-XETG00171__0018220__Region_2__20240207__003742/'\n",
    "output_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_2__20240207__003742.tif'\n",
    "\n",
    "mipurl = input_path + '/morphology_mip.ome.tif'\n",
    "\n",
    "# Assuming `input_path`, `output_path`, `factor`, and `image_size_original` are defined earlier in your script\n",
    "data_file_path = input_path + 'transcripts.csv'  # Ensure `input_path` ends with a slash (/)\n",
    "output_image_path = output_path  # Ensure this variable is correctly defined\n",
    "factor = 0.2125  # Scaling factor to convert pixel into micrometer\n",
    "with tifffile.TiffFile(mipurl) as tif:\n",
    "    image_size_original=tif.series[0].levels[0].shape\n",
    "\n",
    "# Adjust `image_size_original` function or variable appropriately to get original image size\n",
    "image_size = (int(image_size_original[1]*factor), int(image_size_original[0]*factor))\n",
    "\n",
    "\n",
    "# Read the spatial sequencing data\n",
    "data = pd.read_csv(data_file_path)\n",
    "\n",
    "# Create a black image\n",
    "image = Image.new('RGB', image_size, 'black')\n",
    "draw = ImageDraw.Draw(image)\n",
    "\n",
    "# Scale the x_location and y_location by 'factor'\n",
    "data['x_location_scaled'] = data['x_location'] \n",
    "data['y_location_scaled'] = data['y_location'] \n",
    "\n",
    "# Plot the points for the specified genes with qv > 20\n",
    "for index, row in data.iterrows():\n",
    "    if row['qv'] > 20 and row['feature_name'] in genes_to_plot:\n",
    "        color = genes_to_plot[row['feature_name']]\n",
    "        # Plot the gene\n",
    "        draw.point((row['x_location_scaled'], row['y_location_scaled']), fill=color)\n",
    "\n",
    "# Save the image\n",
    "image.save(output_image_path, format='TIFF')\n",
    "\n",
    "print(f\"Image saved to {output_image_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dad2a533",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run for DRG3\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageDraw\n",
    "import tifffile\n",
    "#put in the input folder for the tissue data and the output path to save the transcript image\n",
    "input_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/Xenium_Data/output-XETG00171__0018220__Region_3__20240207__003742/'\n",
    "output_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_3__20240207__003742.tif'\n",
    "\n",
    "mipurl = input_path + '/morphology_mip.ome.tif'\n",
    "\n",
    "# Assuming `input_path`, `output_path`, `factor`, and `image_size_original` are defined earlier in your script\n",
    "data_file_path = input_path + 'transcripts.csv'  # Ensure `input_path` ends with a slash (/)\n",
    "output_image_path = output_path  # Ensure this variable is correctly defined\n",
    "factor = 0.2125  # Scaling factor to convert pixel into micrometer\n",
    "with tifffile.TiffFile(mipurl) as tif:\n",
    "    image_size_original=tif.series[0].levels[0].shape\n",
    "\n",
    "# Adjust `image_size_original` function or variable appropriately to get original image size\n",
    "image_size = (int(image_size_original[1]*factor), int(image_size_original[0]*factor))\n",
    "\n",
    "\n",
    "# Read the spatial sequencing data\n",
    "data = pd.read_csv(data_file_path)\n",
    "\n",
    "# Create a black image\n",
    "image = Image.new('RGB', image_size, 'black')\n",
    "draw = ImageDraw.Draw(image)\n",
    "\n",
    "# Scale the x_location and y_location by 'factor'\n",
    "data['x_location_scaled'] = data['x_location'] \n",
    "data['y_location_scaled'] = data['y_location'] \n",
    "\n",
    "# Plot the points for the specified genes with qv > 20\n",
    "for index, row in data.iterrows():\n",
    "    if row['qv'] > 20 and row['feature_name'] in genes_to_plot:\n",
    "        color = genes_to_plot[row['feature_name']]\n",
    "        # Plot the gene\n",
    "        draw.point((row['x_location_scaled'], row['y_location_scaled']), fill=color)\n",
    "\n",
    "# Save the image\n",
    "image.save(output_image_path, format='TIFF')\n",
    "\n",
    "print(f\"Image saved to {output_image_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "201aedb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run for DRG4\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageDraw\n",
    "import tifffile\n",
    "#put in the input folder for the tissue data and the output path to save the transcript image\n",
    "input_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/Xenium_Data/output-XETG00171__0018220__Region_4__20240207__003742/'\n",
    "output_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_4__20240207__003742.tif'\n",
    "\n",
    "mipurl = input_path + '/morphology_mip.ome.tif'\n",
    "\n",
    "# Assuming `input_path`, `output_path`, `factor`, and `image_size_original` are defined earlier in your script\n",
    "data_file_path = input_path + 'transcripts.csv'  # Ensure `input_path` ends with a slash (/)\n",
    "output_image_path = output_path  # Ensure this variable is correctly defined\n",
    "factor = 0.2125  # Scaling factor to convert pixel into micrometer\n",
    "with tifffile.TiffFile(mipurl) as tif:\n",
    "    image_size_original=tif.series[0].levels[0].shape\n",
    "\n",
    "# Adjust `image_size_original` function or variable appropriately to get original image size\n",
    "image_size = (int(image_size_original[1]*factor), int(image_size_original[0]*factor))\n",
    "\n",
    "\n",
    "# Read the spatial sequencing data\n",
    "data = pd.read_csv(data_file_path)\n",
    "\n",
    "# Create a black image\n",
    "image = Image.new('RGB', image_size, 'black')\n",
    "draw = ImageDraw.Draw(image)\n",
    "\n",
    "# Scale the x_location and y_location by 'factor'\n",
    "data['x_location_scaled'] = data['x_location'] \n",
    "data['y_location_scaled'] = data['y_location'] \n",
    "\n",
    "# Plot the points for the specified genes with qv > 20\n",
    "for index, row in data.iterrows():\n",
    "    if row['qv'] > 20 and row['feature_name'] in genes_to_plot:\n",
    "        color = genes_to_plot[row['feature_name']]\n",
    "        # Plot the gene\n",
    "        draw.point((row['x_location_scaled'], row['y_location_scaled']), fill=color)\n",
    "\n",
    "# Save the image\n",
    "image.save(output_image_path, format='TIFF')\n",
    "\n",
    "print(f\"Image saved to {output_image_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2644a158",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run for TG1\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageDraw\n",
    "import tifffile\n",
    "#put in the input folder for the tissue data and the output path to save the transcript image\n",
    "input_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/Xenium_Data/output-XETG00171__0018212__Region_1__20240207__003742/'\n",
    "output_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcript_images/output-XETG00171__0018212__Region_1__20240207__003742.tif'\n",
    "\n",
    "mipurl = input_path + '/morphology_mip.ome.tif'\n",
    "\n",
    "# Assuming `input_path`, `output_path`, `factor`, and `image_size_original` are defined earlier in your script\n",
    "data_file_path = input_path + 'transcripts.csv'  # Ensure `input_path` ends with a slash (/)\n",
    "output_image_path = output_path  # Ensure this variable is correctly defined\n",
    "factor = 0.2125  # Scaling factor to convert pixel into micrometer\n",
    "with tifffile.TiffFile(mipurl) as tif:\n",
    "    image_size_original=tif.series[0].levels[0].shape\n",
    "\n",
    "# Adjust `image_size_original` function or variable appropriately to get original image size\n",
    "image_size = (int(image_size_original[1]*factor), int(image_size_original[0]*factor))\n",
    "\n",
    "\n",
    "# Read the spatial sequencing data\n",
    "data = pd.read_csv(data_file_path)\n",
    "\n",
    "# Create a black image\n",
    "image = Image.new('RGB', image_size, 'black')\n",
    "draw = ImageDraw.Draw(image)\n",
    "\n",
    "# Scale the x_location and y_location by 'factor'\n",
    "data['x_location_scaled'] = data['x_location'] \n",
    "data['y_location_scaled'] = data['y_location'] \n",
    "\n",
    "# Plot the points for the specified genes with qv > 20\n",
    "for index, row in data.iterrows():\n",
    "    if row['qv'] > 20 and row['feature_name'] in genes_to_plot:\n",
    "        color = genes_to_plot[row['feature_name']]\n",
    "        # Plot the gene\n",
    "        draw.point((row['x_location_scaled'], row['y_location_scaled']), fill=color)\n",
    "\n",
    "# Save the image\n",
    "image.save(output_image_path, format='TIFF')\n",
    "\n",
    "print(f\"Image saved to {output_image_path}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
