{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8828f6cf",
   "metadata": {},
   "source": [
    "# Creating transcripts from Xenium Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "87aef1c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pixel_value = [(255, 0 ,0), (0, 255, 0)] \n",
    "list_of_genes=['UCHL1', 'FABP7'] \n",
    "genes_to_plot = {}\n",
    "for i in range(len(pixel_value)):\n",
    "    genes_to_plot[list_of_genes[i]]= pixel_value[i]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "91f37b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ImageDescription: <OME xmlns=\"http://www.openmicroscopy.org/Schemas/OME/2016-06\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" Creator=\"tifffile.py 2022.10.10\" UUID=\"urn:uuid:f6b24040-c570-11ee-a73b-3cecefe152c4\" xsi:schemaLocation=\"http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd\">\n",
      "    <Plate ID=\"Plate:1\" WellOriginX=\"-0.0\" WellOriginXUnit=\"µm\" WellOriginY=\"-0.0\" WellOriginYUnit=\"µm\" />\n",
      "    <Instrument ID=\"Instrument:1\">\n",
      "        <Microscope Manufacturer=\"10x Genomics\" Model=\"Xenium\" />\n",
      "    </Instrument>\n",
      "    <Image ID=\"Image:0\" Name=\"Image0\">\n",
      "        <InstrumentRef ID=\"Instrument:1\" />\n",
      "        <Pixels DimensionOrder=\"XYZCT\" ID=\"Pixels:0\" SizeC=\"1\" SizeT=\"1\" SizeX=\"17170\" SizeY=\"27295\" SizeZ=\"13\" Type=\"uint16\" PhysicalSizeX=\"0.2125\" PhysicalSizeY=\"0.2125\" PhysicalSizeZ=\"3.0\">\n",
      "            <Channel ID=\"Channel:0:0\" Name=\"DAPI\" SamplesPerPixel=\"1\" />\n",
      "            <TiffData PlaneCount=\"13\" />\n",
      "        </Pixels>\n",
      "    </Image>\n",
      "</OME>\n",
      "0 (13, 27295, 17170)\n",
      "1 (13, 13647, 8585)\n",
      "2 (13, 6823, 4292)\n",
      "3 (13, 3411, 2146)\n",
      "4 (13, 1705, 1073)\n",
      "5 (13, 852, 536)\n",
      "6 (13, 426, 268)\n",
      "passed\n",
      "Image saved to /media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_4__20240207__003742.tif\n",
      "ImageDescription: <OME xmlns=\"http://www.openmicroscopy.org/Schemas/OME/2016-06\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" Creator=\"tifffile.py 2022.10.10\" UUID=\"urn:uuid:3b745864-c56f-11ee-91e5-3cecefe152c4\" xsi:schemaLocation=\"http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd\">\n",
      "    <Plate ID=\"Plate:1\" WellOriginX=\"-0.0\" WellOriginXUnit=\"µm\" WellOriginY=\"-0.0\" WellOriginYUnit=\"µm\" />\n",
      "    <Instrument ID=\"Instrument:1\">\n",
      "        <Microscope Manufacturer=\"10x Genomics\" Model=\"Xenium\" />\n",
      "    </Instrument>\n",
      "    <Image ID=\"Image:0\" Name=\"Image0\">\n",
      "        <InstrumentRef ID=\"Instrument:1\" />\n",
      "        <Pixels DimensionOrder=\"XYZCT\" ID=\"Pixels:0\" SizeC=\"1\" SizeT=\"1\" SizeX=\"14330\" SizeY=\"20501\" SizeZ=\"12\" Type=\"uint16\" PhysicalSizeX=\"0.2125\" PhysicalSizeY=\"0.2125\" PhysicalSizeZ=\"3.0\">\n",
      "            <Channel ID=\"Channel:0:0\" Name=\"DAPI\" SamplesPerPixel=\"1\" />\n",
      "            <TiffData PlaneCount=\"12\" />\n",
      "        </Pixels>\n",
      "    </Image>\n",
      "</OME>\n",
      "0 (12, 20501, 14330)\n",
      "1 (12, 10250, 7165)\n",
      "2 (12, 5125, 3582)\n",
      "3 (12, 2562, 1791)\n",
      "4 (12, 1281, 895)\n",
      "5 (12, 640, 447)\n",
      "6 (12, 320, 223)\n",
      "passed\n",
      "Image saved to /media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_3__20240207__003742.tif\n",
      "ImageDescription: <OME xmlns=\"http://www.openmicroscopy.org/Schemas/OME/2016-06\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" Creator=\"tifffile.py 2022.10.10\" UUID=\"urn:uuid:ccf9062e-c56d-11ee-be36-3cecefe152c4\" xsi:schemaLocation=\"http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd\">\n",
      "    <Plate ID=\"Plate:1\" WellOriginX=\"-0.0\" WellOriginXUnit=\"µm\" WellOriginY=\"-0.0\" WellOriginYUnit=\"µm\" />\n",
      "    <Instrument ID=\"Instrument:1\">\n",
      "        <Microscope Manufacturer=\"10x Genomics\" Model=\"Xenium\" />\n",
      "    </Instrument>\n",
      "    <Image ID=\"Image:0\" Name=\"Image0\">\n",
      "        <InstrumentRef ID=\"Instrument:1\" />\n",
      "        <Pixels DimensionOrder=\"XYZCT\" ID=\"Pixels:0\" SizeC=\"1\" SizeT=\"1\" SizeX=\"22814\" SizeY=\"10334\" SizeZ=\"12\" Type=\"uint16\" PhysicalSizeX=\"0.2125\" PhysicalSizeY=\"0.2125\" PhysicalSizeZ=\"3.0\">\n",
      "            <Channel ID=\"Channel:0:0\" Name=\"DAPI\" SamplesPerPixel=\"1\" />\n",
      "            <TiffData PlaneCount=\"12\" />\n",
      "        </Pixels>\n",
      "    </Image>\n",
      "</OME>\n",
      "0 (12, 10334, 22814)\n",
      "1 (12, 5167, 11407)\n",
      "2 (12, 2583, 5703)\n",
      "3 (12, 1291, 2851)\n",
      "4 (12, 645, 1425)\n",
      "5 (12, 322, 712)\n",
      "6 (12, 161, 356)\n",
      "passed\n",
      "Image saved to /media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_2__20240207__003742.tif\n",
      "ImageDescription: <OME xmlns=\"http://www.openmicroscopy.org/Schemas/OME/2016-06\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" Creator=\"tifffile.py 2022.10.10\" UUID=\"urn:uuid:8f39bb86-c56c-11ee-a387-3cecefe152c4\" xsi:schemaLocation=\"http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd\">\n",
      "    <Plate ID=\"Plate:1\" WellOriginX=\"-0.0\" WellOriginXUnit=\"µm\" WellOriginY=\"-0.0\" WellOriginYUnit=\"µm\" />\n",
      "    <Instrument ID=\"Instrument:1\">\n",
      "        <Microscope Manufacturer=\"10x Genomics\" Model=\"Xenium\" />\n",
      "    </Instrument>\n",
      "    <Image ID=\"Image:0\" Name=\"Image0\">\n",
      "        <InstrumentRef ID=\"Instrument:1\" />\n",
      "        <Pixels DimensionOrder=\"XYZCT\" ID=\"Pixels:0\" SizeC=\"1\" SizeT=\"1\" SizeX=\"34174\" SizeY=\"27280\" SizeZ=\"14\" Type=\"uint16\" PhysicalSizeX=\"0.2125\" PhysicalSizeY=\"0.2125\" PhysicalSizeZ=\"3.0\">\n",
      "            <Channel ID=\"Channel:0:0\" Name=\"DAPI\" SamplesPerPixel=\"1\" />\n",
      "            <TiffData PlaneCount=\"14\" />\n",
      "        </Pixels>\n",
      "    </Image>\n",
      "</OME>\n",
      "0 (14, 27280, 34174)\n",
      "1 (14, 13640, 17087)\n",
      "2 (14, 6820, 8543)\n",
      "3 (14, 3410, 4271)\n",
      "4 (14, 1705, 2135)\n",
      "5 (14, 852, 1067)\n",
      "6 (14, 426, 533)\n",
      "7 (14, 213, 266)\n",
      "passed\n",
      "Image saved to /media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018220__Region_1__20240207__003742.tif\n",
      "ImageDescription: <OME xmlns=\"http://www.openmicroscopy.org/Schemas/OME/2016-06\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" Creator=\"tifffile.py 2022.10.10\" UUID=\"urn:uuid:2103ddb0-c56a-11ee-b31f-3cecefe152c4\" xsi:schemaLocation=\"http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd\">\n",
      "    <Plate ID=\"Plate:1\" WellOriginX=\"-0.0\" WellOriginXUnit=\"µm\" WellOriginY=\"-0.0\" WellOriginYUnit=\"µm\" />\n",
      "    <Instrument ID=\"Instrument:1\">\n",
      "        <Microscope Manufacturer=\"10x Genomics\" Model=\"Xenium\" />\n",
      "    </Instrument>\n",
      "    <Image ID=\"Image:0\" Name=\"Image0\">\n",
      "        <InstrumentRef ID=\"Instrument:1\" />\n",
      "        <Pixels DimensionOrder=\"XYZCT\" ID=\"Pixels:0\" SizeC=\"1\" SizeT=\"1\" SizeX=\"42698\" SizeY=\"64642\" SizeZ=\"17\" Type=\"uint16\" PhysicalSizeX=\"0.2125\" PhysicalSizeY=\"0.2125\" PhysicalSizeZ=\"3.0\">\n",
      "            <Channel ID=\"Channel:0:0\" Name=\"DAPI\" SamplesPerPixel=\"1\" />\n",
      "            <TiffData PlaneCount=\"17\" />\n",
      "        </Pixels>\n",
      "    </Image>\n",
      "</OME>\n",
      "0 (17, 64642, 42698)\n",
      "1 (17, 32321, 21349)\n",
      "2 (17, 16160, 10674)\n",
      "3 (17, 8080, 5337)\n",
      "4 (17, 4040, 2668)\n",
      "5 (17, 2020, 1334)\n",
      "6 (17, 1010, 667)\n",
      "7 (17, 505, 333)\n",
      "passed\n",
      "Image saved to /media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/output-XETG00171__0018212__Region_1__20240207__003742.tif\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import tifffile\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageDraw\n",
    "folder_path = '/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts'\n",
    "os.makedirs(folder_path, exist_ok=True) \n",
    "file_list = os.listdir('/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/Xenium_Data')\n",
    "file_list = sorted(file_list, reverse=True)\n",
    "for i in range(len(file_list)):\n",
    "    input_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/Xenium_segementation/Xenium_output/'+file_list[i]\n",
    "    output_path ='/media/yhs/5596744f-db7c-442f-9235-d0c9d50c0a6b/export/transcripts/'+file_list[i]+'.tif'\n",
    "    mipurl = input_path + '/morphology.ome.tif'\n",
    "\n",
    "    with tifffile.TiffFile(mipurl) as tif:\n",
    "        for tag in tif.pages[0].tags.values():\n",
    "            if tag.name == 'ImageDescription':\n",
    "                print(tag.name+\":\", tag.value)\n",
    "    with tifffile.TiffFile(mipurl) as tif:\n",
    "        image_size_original=tif.series[0].levels[0].shape\n",
    "        for i in range(len(tif.series[0].levels)):\n",
    "            print(i, tif.series[0].levels[i].shape)\n",
    "\n",
    "    # Assuming `input_path`, `output_path`, `factor`, and `image_size_original` are defined earlier in your script\n",
    "    data_file_path = input_path + '/transcripts.csv'  # Ensure `input_path` ends with a slash (/)\n",
    "    output_image_path = output_path  # Ensure this variable is correctly defined\n",
    "    factor = 0.2125  # Scaling factor to convert pixel into micrometer\n",
    "\n",
    "    # Adjust `image_size_original` function or variable appropriately to get original image size\n",
    "    image_size = (int(image_size_original[2]*factor), int(image_size_original[1]*factor))\n",
    "\n",
    "    # Read the spatial sequencing data\n",
    "    data = pd.read_csv(data_file_path)\n",
    "\n",
    "    # Create a black image\n",
    "    image = Image.new('RGB', image_size, 'black')\n",
    "    draw = ImageDraw.Draw(image)\n",
    "\n",
    "    # Scale the x_location and y_location by 'factor'\n",
    "    data['x_location_scaled'] = data['x_location'] \n",
    "    data['y_location_scaled'] = data['y_location'] \n",
    "    print(\"passed\")\n",
    "\n",
    "    # Plot the points for the specified genes with qv > 20\n",
    "    for index, row in data.iterrows():\n",
    "        if row['qv'] > 20 and row['feature_name'] in genes_to_plot:\n",
    "            color = genes_to_plot[row['feature_name']]\n",
    "            # Plot the gene\n",
    "            draw.point((row['x_location_scaled'], row['y_location_scaled']), fill=color)\n",
    "\n",
    "    # Save the image\n",
    "    image.save(output_image_path, format='TIFF')\n",
    "\n",
    "    print(f\"Image saved to {output_image_path}\")\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
